rust/lance/src/io/commit.rs

```diff
@@ -200,6 +201,73 @@ async fn migrate_manifest(
     Ok(())
 }
 
+/// Fix schema in case of duplicate field ids.
+///
+/// See test dataset v0.10.5/corrupt_schema
+fn fix_schema(manifest: &mut Manifest) -> Result<()> {
+    let mut field_id_seed = manifest.max_field_id().unwrap_or(-1) + 1;
+    let mut seen_fields = HashSet::new();
+    // This map is for transforming the data files' field lists
+    let mut file_field_mapping: HashMap<(usize, usize), i32> = HashMap::new();
+    // This map is for transforming the schema
+    let mut old_field_id_mapping: HashMap<i32, i32> = HashMap::new();
+    let field_ids = if let Some(fragment) = manifest.fragments.first() {
+        fragment
+            .files
+            .iter()
+            .enumerate()
+            .flat_map(|(f_pos, f)| std::iter::repeat(f_pos).zip(f.fields.iter().enumerate()))
+    } else {
+        return Ok(());
+    };
+    for (file_pos, (field_pos, field_id)) in field_ids {
+        if !seen_fields.insert(field_id) {
+            let new_field_id = field_id_seed;
+            field_id_seed += 1;
+
+            file_field_mapping.insert((file_pos, field_pos), new_field_id);
+            old_field_id_mapping.insert(*field_id, new_field_id);
+        }
+    }
+
+    if file_field_mapping.is_empty() {
+        return Ok(());
+    }
+
+    let mut fragments = manifest.fragments.as_ref().clone();
+
+    // Apply mapping to fragment files list
+    for fragment in fragments.iter_mut() {
+        for ((file_pos, field_pos), new_field_id) in &file_field_mapping {
+            fragment.files[*file_pos].fields[*field_pos] = *new_field_id;
+        }
+    }
+
+    // Apply mapping to the schema
+    for (old_field_id, new_field_id) in &old_field_id_mapping {
+        let field = manifest.schema.mut_field_by_id(*old_field_id).unwrap();
+        field.id = *new_field_id;
+    }
```

Comment:
How does this part work.  If the data files had a duplicate id we have something like...

```
Fragment 0:
  Data File 0:
    Field 0
    Field 1
  Data File 1:
    Field 0
```

I understand the above loop (replace (file-1,field-0) with field id 2) but how can we fix the schema? Isn't the schema:

```
Schema:
  field 0
  field 1
  field 0
```

If we run through this loop aren't we replacing the first instance of `field 0` with `field 2`?  Wouldn't we want to replace the last instance of `field 0`?

Reply:
The schema will never have duplicate ids, since we've been using the `max_field_id` from the schema thus far. The case is more like:

```
Schema:
  field 1
  field 0
```

Where `field 0` is loading the one from `Data File 0` when it should be loading from `Data File 1`. Therefore, we map it to `field 2` so that there's no conflict.

Reply:
Ah, I see, this only fixes schemas that have corrupted by the specific add/remove/add-back process.  So the field ids we are mapping from are field ids that shouldn't exist in the schema at all.